sycl/source/detail/scheduler/commands.cpp

```diff
@@ -156,58 +157,72 @@ getPiEvents(const std::vector<EventImplPtr> &EventImpls) {
-// Method prepares PI event's from list sycl::event's
-std::vector<EventImplPtr> Command::prepareEvents(ContextImplPtr Context) {
-  std::vector<EventImplPtr> Result;
-  std::vector<EventImplPtr> GlueEvents;
-  for (EventImplPtr &DepEvent : MDepsEvents) {
-    // Async work is not supported for host device.
-    if (DepEvent->is_host()) {
-      DepEvent->waitInternal();
-      continue;
+  void waitForEvents() const {
+    std::map<const detail::plugin *, std::vector<EventImplPtr>>
+        RequiredEventsPerPlugin;
+
+    for (const EventImplPtr &Event : MDepEvents) {
+      const detail::plugin &Plugin = Event->getPlugin();
+      RequiredEventsPerPlugin[&Plugin].push_back(Event);
     }
-    // The event handle can be null in case of, for example, alloca command,
-    // which is currently synchrounious, so don't generate OpenCL event.
-    if (DepEvent->getHandleRef() == nullptr) {
-      continue;
+
+    // wait for dependency events
+    // FIXME introduce a more sophisticated wait mechanism
+    for (auto &PluginWithEvents : RequiredEventsPerPlugin) {
+      std::vector<RT::PiEvent> RawEvents = getPiEvents(PluginWithEvents.second);
+      PluginWithEvents.first->call<PiApiKind::piEventsWait>(RawEvents.size(),
+                                                            RawEvents.data());
     }
-    ContextImplPtr DepEventContext = DepEvent->getContextImpl();
-
-    // If contexts don't match - connect them using user event
-    if (DepEventContext != Context && !Context->is_host()) {
-      EventImplPtr GlueEvent(new detail::event_impl());
-      GlueEvent->setContextImpl(Context);
-      EventImplPtr *GlueEventCopy =
-          new EventImplPtr(GlueEvent); // To increase the reference count by 1.
-
-      RT::PiEvent &GlueEventHandle = GlueEvent->getHandleRef();
-      auto Plugin = Context->getPlugin();
-      auto DepPlugin = DepEventContext->getPlugin();
-      // Add an event on the current context that
-      // is triggered when the DepEvent is complete
-      Plugin.call<PiApiKind::piEventCreate>(Context->getHandleRef(),
-                                            &GlueEventHandle);
-
-      DepPlugin.call<PiApiKind::piEventSetCallback>(
-          DepEvent->getHandleRef(), PI_EVENT_COMPLETE, EventCompletionClbk,
-          /*void *data=*/(GlueEventCopy));
-      GlueEvents.push_back(GlueEvent);
-      Result.push_back(std::move(GlueEvent));
-      continue;
+  }
+
+public:
+  DispatchHostTask(std::vector<EventImplPtr> DepEvents, CGHostTask *HostTask,
+                   std::vector<DepDesc> Deps, EventImplPtr SelfEvent)
+      : MDepEvents(std::move(DepEvents)), MHostTask{HostTask},
+        MDeps(std::move(Deps)), MSelfEvent(std::move(SelfEvent)) {}
+
+  void operator()() const {
+    waitForEvents();
+
+    // we're ready to call the user-defined lambda now
+    MHostTask->MHostTask->call();
+
+    // update self-event status
+    if (MSelfEvent->is_host()) {
+      // TODO
+      fprintf(stderr, "Gonna enqueue smth here\n");
```

Comment:
Have to enqueu smth here. I believe this is the cause of race condition for now.

Reply:
Done.